clang/lib/CConv/CheckedRegions.cpp

```diff
@@ -28,18 +30,112 @@ using namespace clang;
 
 bool CheckedRegionAdder::VisitCompoundStmt(CompoundStmt *S) { 
   llvm::FoldingSetNodeID Id;
+  ast_type_traits::DynTypedNode DTN = ast_type_traits::DynTypedNode::create(*S);
+
   S->Profile(Id, *Context, true);
   switch (Map[Id]) {
-    case IS_UNCHECKED: return true;
-    case IS_CHECKED:   
-                        auto Loc = S->getBeginLoc();
-                        Writer.InsertTextBefore(Loc, "_Checked ");
-                        return false;
+    case IS_UNCHECKED:
+      if(isParentChecked(DTN) && !isFunctionBody(S)) {
+        auto Loc = S->getBeginLoc();
+        Writer.InsertTextBefore(Loc, "_Unchecked ");
+      }
+      break;
+    case IS_CHECKED:
+      if(!isParentChecked(DTN)) {
+        auto Loc = S->getBeginLoc();
+        Writer.InsertTextBefore(Loc, "_Checked ");
+      }
+      break;
+    default: llvm_unreachable("Bad flag in CheckedRegionAdder");
   }
 
-  llvm_unreachable("Bad flag in CheckedRegionAdder");
+  for(const auto C : S->children()) {
+    CheckedRegionAdder SV(Context, Writer, Map);
+    SV.TraverseStmt(C);
+  }
```

Comment:
Why manually traverse the children and return `false` rather than return `true` to let the parent visitor do it? I'm also surprised that you are looking at the children *after* you have made a decision about this, the compound statement.

Reply:
This visitor is the second pass, so the only information that needs to be checked is what decision was made for the node above.
On the child traversal I made the change as suggested.

Reply:
Got it.